### PR TITLE
Implements the new activegate section

### PR DESCRIPTION
**Pull request by 0sewa0**
Implements the new activeGate section where each capability we want to turn on is put into a list, but the rest of the config they share, and capabilities get deployed in 1 pod in 1 container. If the customer wants the capabilities to have their on pods then multiple Dynakube need to be created.

### CODE REVIEW

**Review by aorcholski — CHANGES_REQUESTED** [review]
"make deploy" command does not create serviceaccount/dynatrace-multi so statfulset reports an error.
dynakube object :
```
apiVersion: dynatrace.com/v1beta1                                                                                                                                                                                  
kind: DynaKube                                                                                                                                                                                                     
metadata:                                                                                                                                                                                                          
  name: dk                                                                                                                                                                                                         
  namespace: dynatrace                                                                                                                                                                                             
spec:                                                                                                                                                                                                              
  apiUrl: https://...
  activeGate:                                                                                                                                                                                                      
    capabilities:                                                                                                                                                                                                  
      - "routing" 
```

File: api/v1beta1/activegate_types.go [hunk 1]
```diff
@@ -4,19 +4,62 @@ import (
...
-const (
-	Routing    ActiveGateCapability = "routing"
-	KubeMon    ActiveGateCapability = "kubernetes-monitoring"
-	DataIngest ActiveGateCapability = "data-ingest"
+type ActiveGateCapability struct {
+
+	// The name of the capability known by the user, mainly used in the CR
+	DisplayName CapabilityDisplayName
+
+	// The name of used for marking the pod for given capability
```
Comment (chrismuellner) [inline api/v1beta1/activegate_types.go:14]:
```suggestion
	// The name used for marking the pod for given capability
```

File: api/v1beta1/activegate_types.go [hunk 2]
```diff
@@ -4,19 +4,62 @@ import (
...
+	}
 )
 
 type ActiveGateSpec struct {
 
 	// Activegate capabilities enabled (routing, kubernetes-monitoring, data-ingest)
-	Capabilities []ActiveGateCapability `json:"capabilities,omitempty"`
+	Capabilities []CapabilityDisplayName `json:"capabilities,omitempty"`
+
+	// Amount of replicas for your DynaKube
```
Comment (chrismuellner) [inline api/v1beta1/activegate_types.go:46]:
```suggestion
	// Amount of replicas for the ActiveGate
```

File: api/v1beta1/dynakube_conversion.go [hunk 3]
```diff
@@ -183,8 +183,6 @@ func (dst *DynaKube) ConvertFrom(srcRaw conversion.Hub) error {
 }
 
 func convertToDeprecatedActiveGateCapability(dst *CapabilityProperties, src *v1alpha1.CapabilityProperties) {
-	dst.Enabled = true
```
Comment (chrismuellner) [inline api/v1beta1/dynakube_conversion.go:186]: Is this not needed?

File: config/crd/default/patches/webhook_in_dynakubes.yaml [hunk 4]
```diff
@@ -13,4 +13,4 @@ spec:
           name: dynatrace-webhook
           path: /convert
       conversionReviewVersions:
-      - v1
+      - v1beta1
```
Comment (chrismuellner) [inline config/crd/default/patches/webhook_in_dynakubes.yaml:16]: does not belong in this PR

File: config/common/activegates/serviceaccount-activegates.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: dynatrace-activegates
```
Comment (chrismuellner) [inline config/common/activegates/serviceaccount-activegates.yaml:4]: Wouldn't `dynatrace-activegate` fit better (similar to the other Service Accounts)?

File: controllers/activegate/capability/capability.go [hunk 6]
```diff
@@ -57,12 +73,12 @@ func (c *capabilityBase) GetConfiguration() Configuration {
 	return c.Configuration
 }
 
-func (c *capabilityBase) GetModuleName() string {
-	return c.moduleName
+func (c *capabilityBase) GetShortName() string {
```
Comment (chrismuellner) [inline controllers/activegate/capability/capability.go:76]:
```suggestion
func (c *capabilityBase) ShortName() string {
```

File: controllers/activegate/capability/capability.go [hunk 7]
```diff
@@ -57,12 +73,12 @@ func (c *capabilityBase) GetConfiguration() Configuration {
...
+	return c.shortName
 }
 
-func (c *capabilityBase) GetCapabilityName() string {
-	return c.capabilityName
+func (c *capabilityBase) GetArgName() string {
```
Comment (chrismuellner) [inline controllers/activegate/capability/capability.go:80]:
```suggestion
func (c *capabilityBase) ArgName() string {
```

**Review by chrismuellner — CHANGES_REQUESTED** [review]

File: config/common/activegates/serviceaccount-activegates.yaml (hunk 5 above)
Comment (0sewa0) [reply config/common/activegates/serviceaccount-activegates.yaml:4]: but this can have multiple activegates inside :)

File: controllers/activegate/capability/capability.go (hunk 6 above)
Comment (0sewa0) [reply controllers/activegate/capability/capability.go:76]: then I should redo all the other functions in the interface, which will result in more files changed, if you are up for it then I will do it 😛

File: config/common/activegates/serviceaccount-activegates.yaml (hunk 5 above)
Comment (chrismuellner) [reply config/common/activegates/serviceaccount-activegates.yaml:4]: isn't it still one activegate that has multiple capabilities?

File: config/crd/default/patches/webhook_in_dynakubes.yaml (hunk 4 above)
Comment (0sewa0) [reply config/crd/default/patches/webhook_in_dynakubes.yaml:16]: its already in master, so removing it here does nothing

File: config/common/activegates/serviceaccount-activegates.yaml (hunk 5 above)
Comment (0sewa0) [reply config/common/activegates/serviceaccount-activegates.yaml:4]: 😞 I hate naming things

File: api/v1beta1/dynakube_conversion.go (hunk 3 above)
Comment (0sewa0) [reply api/v1beta1/dynakube_conversion.go:186]: Its not part of CapabilityProperties, moved it out before the func call

**Review by aorcholski — APPROVED** [review]

**Review by chrismuellner — APPROVED** [review]

File: controllers/activegate/capability/capability.go [hunk 8]
```diff
@@ -30,169 +40,235 @@ type Configuration struct {
...
 		},
 	}
+	if dk == nil || !dk.ActiveGateMode() {
+		return &mc
+	}
+	mc.enabled = true
+	mc.properties = &dk.Spec.ActiveGate.CapabilityProperties
+	capabilityNames := []string{}
+	for _, capName := range dk.Spec.ActiveGate.Capabilities {
+		cap := activeGateCapabilities[capName]()
```
Comment (aorcholski) [inline controllers/activegate/capability/capability.go:155]: There is no protection against invalid DisplayName in [capName]. Operator dies:
```pod/dynatrace-operator-7b66c9cdf9-8czlm   0/1     CrashLoopBackOff   5 (61s ago)   9m5s```

```E1001 11:21:16.848245       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)```

I added 'dummy' capability using 'kubectl edit mydynakube' command:

```
spec:
  activeGate:
    capabilities:
    - kubernetes-monitoring
    - dummy
```

**Review by aorcholski — CHANGES_REQUESTED** [review]

File: controllers/activegate/capability/capability.go (hunk 8 above)
Comment (chrismuellner) [reply controllers/activegate/capability/capability.go:155]: This will be fixed with an updated validation webhook

Comment (0sewa0) [reply controllers/activegate/capability/capability.go]: yes, the validation webhook will catch this, but I also added a check so in case we want to support this on openshift 3.11 (that doesn't have webhooks) we can do so

File: controllers/activegate/capability/capability.go [hunk 9]
```diff
@@ -152,7 +152,12 @@ func NewMultiCapability(dk *dynatracev1beta1.DynaKube) *MultiCapability {
 	mc.properties = &dk.Spec.ActiveGate.CapabilityProperties
 	capabilityNames := []string{}
 	for _, capName := range dk.Spec.ActiveGate.Capabilities {
-		cap := activeGateCapabilities[capName]()
+		capabilityGenerator, ok := activeGateCapabilities[capName]
+		if !ok {
+			continue
```
Comment (chrismuellner) [inline controllers/activegate/capability/capability.go:157]: I guess ignoring invalid capabilities is fine

**Review by chrismuellner — APPROVED** [review]
